File: src-tauri/src/commands/credentials/discovery.rs
use std::collections::HashMap;
use std::sync::Arc;

use tauri::State;

use crate::engine::discovery::{self, DiscoveredItem};
use crate::error::AppError;
use crate::AppState;
use personas_macros::requires;

/// Resolve a discovery op for a credential and return the discovered items
/// (projects, environments, repos, etc.) as `{value, label, sublabel}` triples.
///
/// Used by the template adoption questionnaire to populate dynamic option
/// lists from real connector data instead of asking users to type identifiers.
#[tauri::command]
#[requires(privileged)]
pub async fn discover_connector_resources(
    state: State<'_, Arc<AppState>>,
    credential_id: String,
    service_type: String,
    operation: String,
    params: Option<HashMap<String, String>>,
) -> Result<Vec<DiscoveredItem>, AppError> {
    discovery::discover_resources(
        &state.db,
        &credential_id,
        &service_type,
        &operation,
        params.unwrap_or_default(),
    )
    .await
}

File: src/api/discovery/discovery.ts
import { invokeWithTimeout as invoke } from "@/lib/tauriInvoke";

/**
 * Single item returned by a connector discovery op — used to populate
 * dynamic option lists in the adoption questionnaire (Sentry projects,
 * codebase names, etc.).
 *
 * Mirrors the Rust `DiscoveredItem` struct in `engine/discovery.rs`.
 */
export interface DiscoveredItem {
  value: string;
  label: string;
  sublabel: string | null;
}

/**
 * Ask the backend to resolve a discovery op against a specific credential
 * and return a list of `{value, label, sublabel}` triples.
 *
 * `params` carries user-supplied values used by chained discovery ops
 * (e.g. a project slug the user already picked, interpolated as
 * `{{param.<key>}}` in the op's path template).
 */
export const discoverConnectorResources = (
  credentialId: string,
  serviceType: string,
  operation: string,
  params?: Record<string, string>,
) =>
  invoke<DiscoveredItem[]>("discover_connector_resources", {
    credentialId,
    serviceType,
    operation,
    params: params ?? {},
  });
